integration-tests/upgradeable-contracts/README.md

```diff
@@ -0,0 +1,36 @@
+# Upgradeable Contracts
+
+There are different ways a contract can be upgraded in ink!
+
+This folder illustrates some of the common and best practices to achieve upgradeability in your contracts.
+
+## [`set-code-hash`](set-code-hash/)
+
+ink! provides an ability to replace the code under the given contract's address.
+This is exactly what `set_code_hash()` function does.
+
+However, developers needs to be mindful of storage compatibility.
+You can read more about storage compatibility on [use.ink](https://use.ink/basics/upgradeable-contracts#replacing-contract-code-with-set_code_hash)
+
+## [Delegator](delegator/)
+
+Delegator patter is based around a low level cross contract call function `delegate_call`.
+It allows a contract to delegate its execution to some on-chain uploaded code.
```

Comment:
I think it's a mistake to present these as two separate ways of upgrading storage - `set_code_hash` is almost always used together with `delegate_call` to migrate the contract's storage after upgrading its code.

Reply:
`delegate_call` doesn't upgrade the contract - i.e. it doesn't change it's logic (one could argue that if all calls delegate then it is but I'd say it's a proxy then) it can only change its data. Hence the only way to upgrade the contract is to:
1. forward to a different contract - proxy pattern
2. replace the code - `set_code_hash`. 


Also, you can only delegate if that logic has been there since the beginning - i.e. since the moment the contract was uploaded. I'd argue that the contract isn't _upgraded_ .

I think the nomenclature is important here b/c if used loosely it will only add to devs confusion.

Reply:
`delegate_call` allows to delegate an execution of logic to another on-chain. If used correctly, you can introduce the upgradeability functionality where delegated code hash can be replaced. The reason why it belongs to upgradeability is because that's how it is implemented in the EVM world - via proxies. 

`set_code_hash` is the "Polkadot" way of upgradability. 

Nonetheless, these nuances should be elaborated in the documentation. The examples just provide a concrete illustration of how these "blocks" can be used individually.